File: ui/src/surface/tunepath.ts
// **The `tune` layer** (T-898, docs/23 §10.6 rule 2): *where the radio has been* — the device's own
// route through frequency, traced on the map like a directions line.
//
// The user's second map-UI principle is that anything with (time × frequency) coordinates is drawn
// on the surface. The front end's own movement has them: at every instant it was tuned somewhere,
// and each retune moved it. With the canvas's axes (X = frequency, Y = time) that draws as a
// **vertical run per dwell** and a **near-horizontal step per retune** — a staircase up the screen
// showing, at a glance, where the survey has actually been.
//
// ## Nothing here derives the tune history
//
// Every vertex is a backend answer (`GET /api/tune-history`, derived by `hk_store::tunepath` from
// the tune records the coverage map is already built on). This file converts vertices into strokes
// inside one pane, through the very `toClip` the tiles are placed with, from inside
// `SurfaceView.frame()` — so a vertex and the tile row at its capture time are placed by one
// mapping on one frame and cannot drift apart (T-388's structural fix). The poll only refreshes the
// records; it never positions anything.
//
// ## One line per front end, labelled by `device_id`
//
// The several-SDRs rule: each radio's route is its own path with its own ink, and the key in the
// layers menu names the device beside its colour ([[tuneKeyEntries]]). `unknown` — a record that
// named no radio — is its own route and never folded into a named one, exactly as `/api/coverage`
// keeps them apart.
//
// ## It is the device's route, not a measurement
//
// The stroke is drawn in its own palette, distinct from every measurement mark, and it is only ever
// a stroke: like the `paths` layer it submits axis-aligned `path-stroke` rectangles through
// [[pathQuads]], so the data pass stays byte-identical with the layer on and off. A gap the backend
// broke the route at is simply two paths, and nothing is drawn between them — the client never
// joins them, because not knowing where the radio was is not something to draw over.

import type { Box } from "./lattice";
import type { OverlayQuad } from "./minimap";
import { pathQuads, type PathStyle } from "./paths";
import type { PaneRect } from "./surface";

const S_TO_NS = 1e9;

/** Stroke thickness of a device route, device px. */
export const TUNE_STROKE_PX = 2;

/**
 * The inks a device route may be drawn in, in assignment order: an Okabe–Ito-derived set, chosen
 * colour-blind safe (T-813's rule) and away from every measurement mark — the confirmed teal, the
 * candidate purple, the selection orange and the `paths` layer's blue — so the radio's own route
 * can never be mistaken for a signal.
 */
export const TUNE_INKS: readonly (readonly [number, number, number, number])[] = [
  [0.84, 0.37, 0.0, 0.9], // vermillion
  [0.94, 0.89, 0.26, 0.9], // yellow
  [0.8, 0.47, 0.65, 0.9], // reddish purple
  [0.0, 0.45, 0.7, 0.9], // dark blue
  [0.35, 0.7, 0.9, 0.9], // sky
];

/** One device's route as the client holds it: ordered vertices at absolute capture time. */
export interface TunePath {
  readonly id: string;
  /** The front end this is the route of (`unknown` for a record that named none). */
  readonly device: string;
  readonly deviceNamed: boolean;
  readonly retunes: number;
  readonly vertices: readonly { readonly tNs: number; readonly fHz: number }[];
}

/**
 * The request for the panes whose `tune` layer is on: one `GET /api/tune-history` over the union of
 * their boxes, or `null` when there is nothing to ask about. A viewport route answers about a
 * viewport, so all four bounds are always sent.
 */
export function tuneHistoryRequest(boxes: readonly Box[]): string | null {
  const ok = boxes.filter((b) => b.f1Hz > b.f0Hz && b.t1Ns > b.t0Ns && b.t0Ns > 0);
  if (ok.length === 0) return null;
  const f0 = Math.max(0, Math.min(...ok.map((b) => b.f0Hz)));
  const f1 = Math.max(...ok.map((b) => b.f1Hz));
  const t0 = Math.min(...ok.map((b) => b.t0Ns)) / S_TO_NS;
  const t1 = Math.max(...ok.map((b) => b.t1Ns)) / S_TO_NS;
  return `/api/tune-history?f_lo=${f0}&f_hi=${f1}&t0=${t0}&t1=${t1}`;
}

/** The wire answer (`docs/api.md` "GET /api/tune-history") → [[TunePath]]s. Anything malformed is
 * dropped rather than drawn: a vertex with no time or frequency has no place on the map. */
export function parseTuneHistory(body: unknown): TunePath[] {
  const list = (body as { paths?: unknown } | null)?.paths;
  if (!Array.isArray(list)) return [];
  const out: TunePath[] = [];
  for (const p of list as Record<string, unknown>[]) {
    if (typeof p?.id !== "string" || typeof p?.device !== "string") continue;
    const vs = Array.isArray(p.vertices) ? (p.vertices as Record<string, unknown>[]) : [];
    const vertices = vs
      .filter((v) => typeof v?.t_s === "number" && typeof v?.f_hz === "number"
        && Number.isFinite(v.t_s) && Number.isFinite(v.f_hz))
      .map((v) => ({ tNs: (v.t_s as number) * S_TO_NS, fHz: v.f_hz as number }));
    if (vertices.length < 2) continue;
    out.push({
      id: p.id,
      device: p.device,
      deviceNamed: p.device_named === true,
      retunes: typeof p.retunes === "number" ? p.retunes : 0,
      vertices,
    });
  }
  return out;
}

/** The devices with a route, in the order the answer named them (stable, so the inks are stable). */
export function tuneDevices(paths: readonly TunePath[]): string[] {
  const seen: string[] = [];
  for (const p of paths) if (!seen.includes(p.device)) seen.push(p.device);
  return seen;
}

/** The ink of one device's route: its position in [[tuneDevices]], wrapped into [[TUNE_INKS]]. */
export function tuneInk(device: string, devices: readonly string[]): readonly [number, number, number, number] {
  const i = devices.indexOf(device);
  return TUNE_INKS[(i < 0 ? 0 : i) % TUNE_INKS.length];
}

/**
 * Stroke every device's route into one pane, each in its own ink, through the tiles' own mapping
 * ([[pathQuads]] does the clipping and the stepping).
 */
export function tuneQuads(
  paths: readonly TunePath[],
  paneBox: Box,
  rect: PaneRect,
  style: PathStyle = {},
): OverlayQuad[] {
  const devices = tuneDevices(paths);
  const out: OverlayQuad[] = [];
  for (const p of paths) {
    out.push(...pathQuads(
      [{ id: p.id, kind: "sweep", vertices: p.vertices }],
      paneBox,
      rect,
      { strokePx: TUNE_STROKE_PX, ...style, rgba: style.rgba ?? tuneInk(p.device, devices) },
    ));
  }
  return out;
}

/** One row of the layer's key: a front end, its ink, and what the line claims. */
export interface TuneKeyEntry {
  readonly key: string;
  readonly label: string;
  readonly note: string;
  readonly rgb: readonly [number, number, number];
}

/**
 * The key for the layers menu: **one row per front end**, so the route on screen is labelled by the
 * radio that drove it. With no route in the window the key says exactly that, rather than showing a
 * swatch for a line nobody can see.
 */
export function tuneKeyEntries(paths: readonly TunePath[]): TuneKeyEntry[] {
  const devices = tuneDevices(paths);
  if (devices.length === 0) {
    return [{
      key: "none",
      label: "No route in view",
      note: "No tune record covers this window — not a claim that the radio was nowhere.",
      rgb: [0.55, 0.58, 0.6],
    }];
  }
  return devices.map((d) => {
    const mine = paths.filter((p) => p.device === d);
    const retunes = mine.reduce((n, p) => n + p.retunes, 0);
    const ink = tuneInk(d, devices);
    const named = mine.every((p) => p.deviceNamed);
    return {
      key: d,
      label: named ? d : `${d} (no device named)`,
      note: `${retunes} retune${retunes === 1 ? "" : "s"} in view`
        + (mine.length > 1 ? `, in ${mine.length} runs — a gap no record covers is not drawn across.` : "."),
      rgb: [ink[0], ink[1], ink[2]] as const,
    };
  });
}

File: crates/hk-store/src/tunepath.rs
//! **The device's own route through frequency, as a traced path** (T-898, docs/23 §10.6 rule 2).
//!
//! The user's second map-UI principle is that anything with (time × frequency) coordinates belongs
//! *on* the map. The radio's own movement has coordinates: at every instant it was tuned somewhere,
//! and each retune moved it. That is a polyline — "like Google Maps' directions line" — and this
//! module derives it.
//!
//! # It is a read of records that already exist, not a new ledger
//!
//! The input is exactly the [`CoverageSpan`]s the coverage map is rasterised from — IQ-ring journal
//! segments, sealed observation-log dwells and sweep visits, and the dwell in flight — each already
//! carrying the tune in force (`center_hz`, `sample_rate_hz`), the interval, and which front end
//! ([`Device`]). Nothing here asks the hardware anything or keeps a second history: if the coverage
//! map says the radio was at 433.92 MHz from t0 to t1, the path goes through (t0, 433.92 MHz).
//!
//! # Three things the derivation has to get right
//!
//! - **One timeline per front end** (the several-SDRs rule). Two radios are two routes; they never
//!   merge into one line, and [`Device::Unknown`] — a record that named no radio — is its own
//!   timeline, never folded into a named one (the same rule `/api/coverage`'s `devices[]` keeps).
//! - **The same tune, seen by several sources, is one leg.** A dwell is usually described three
//!   times over — by the ring journal, by the sealed record, and (at the live edge) by the open
//!   dwell — and a notched window contributes three spans of its own (T-595's analysed sides plus
//!   the DC notch). All of those carry the *same* centre, so they are unioned into one leg rather
//!   than drawn as a stack of coincident lines.
//! - **A gap is not a leg.** Where no record covers an interval, nothing says where the radio was,
//!   and a line drawn across it would be an invented claim (the same honesty the coverage grey
//!   keeps). So a run is **broken** at a gap longer than [`TunePathConfig::join_gap_ns`]: the
//!   device gets several [`TunePath`]s, each labelled with it, and the client draws no ink between
//!   them.
//!
//! # The shape of the line
//!
//! With the canvas's axes (X = frequency, Y = time) a **dwell is a vertical run** — the same centre
//! over an interval — and a **retune is a vertex** at the instant the tune changed, joined to the
//! next dwell's first vertex by a near-horizontal jump. So each leg contributes two vertices, its
//! start and its end, in time order.

use std::collections::BTreeMap;

use hk_model::{TimeRange, Timestamp};

use crate::coverage::{CoverageSpan, Device};

/// The method string served beside a derived path, so an answer names the code that made it.
pub const TUNE_PATH_METHOD: &str = "hk-store/tune-path@1";

/// Longest gap, ns, that two legs may be joined across by default: 2 s.
///
/// A sealed dwell record closes as the retune is commanded and the next opens where it closed, and
/// the ring journal opens a segment on every provenance change, so consecutive legs of a continuous
/// run abut to within a frame. Two seconds is comfortably above that jitter and far below any
/// interval in which the radio could have been somewhere the records do not mention.
pub const DEFAULT_JOIN_GAP_NS: i64 = 2_000_000_000;

/// Most vertices one [`TunePath`] carries. A background sweep visits a hop every few milliseconds,
/// so a wide window's route is genuinely enormous; the newest vertices are kept and the path says
/// it was truncated.
pub const MAX_PATH_VERTICES: usize = 2_000;

/// How [`tune_paths`] joins and bounds.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct TunePathConfig {
    /// Longest gap two legs are joined across; a longer one starts a new path.
    pub join_gap_ns: i64,
    /// Most vertices one path carries (the newest are kept).
    pub max_vertices: usize,
}

impl Default for TunePathConfig {
    fn default() -> Self {
        Self {
            join_gap_ns: DEFAULT_JOIN_GAP_NS,
            max_vertices: MAX_PATH_VERTICES,
        }
    }
}

/// One interval at one tune: the unit a dwell's vertical run is drawn from.
#[derive(Clone, Copy, Debug, PartialEq)]
pub struct TuneLeg {
    /// The interval the front end was at this tune, in absolute capture time.
    pub time: TimeRange,
    /// The tuned RF centre, Hz — the path's frequency coordinate over that interval.
    pub center_hz: f64,
    /// The sample rate in force, Hz (the instantaneous bandwidth around the centre).
    pub sample_rate_hz: f64,
}

/// A vertex of the drawn line: a (capture time, frequency) point.
#[derive(Clone, Copy, Debug, PartialEq)]
pub struct TuneVertex {
    /// Absolute capture time.
    pub t: Timestamp,
    /// The tuned centre then, Hz.
    pub f_hz: f64,
    /// The rate in force, Hz.
    pub sample_rate_hz: f64,
    /// Whether this is the start or the end of its leg.
    pub at: TuneVertexAt,
}

/// Which end of its leg a vertex is.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum TuneVertexAt {
    /// The instant the front end arrived at this tune.
    Start,
    /// The instant it left it (or, for the newest leg, how far the record reaches).
    End,
}

impl TuneVertexAt {
    /// The wire value.
    pub fn as_str(self) -> &'static str {
        match self {
            TuneVertexAt::Start => "start",
            TuneVertexAt::End => "end",
        }
    }
}

/// One front end's continuous route: an ordered run of legs with no unrecorded gap inside it.
#[derive(Clone, Debug, PartialEq)]
pub struct TunePath {
    /// Which front end this is the route of. [`Device::Unknown`] is its own route, never a named
    /// radio's.
    pub device: Device,
    /// The legs, in time order.
    pub legs: Vec<TuneLeg>,
    /// Whether legs were dropped to fit [`TunePathConfig::max_vertices`] (the oldest go first).
    pub truncated: bool,
}

impl TunePath {
    /// The run's time extent.
    pub fn time(&self) -> TimeRange {
        let (a, b) = (
            self.legs.first().map(|l| l.time.start),
            self.legs.last().map(|l| l.time.end),
        );
        match (a, b) {
            (Some(a), Some(b)) => TimeRange::new(a, b),
            _ => TimeRange::new(Timestamp::from_unix_nanos(0), Timestamp::from_unix_nanos(0)),
        }
    }

    /// The lowest and highest centre the route reached, Hz (`None` for an empty route).
    pub fn freq_extent(&self) -> Option<(f64, f64)> {
        let mut it = self.legs.iter().map(|l| l.center_hz);
        let first = it.next()?;
        Some(it.fold((first, first), |(lo, hi), c| (lo.min(c), hi.max(c))))
    }

    /// How many retunes the route contains: a change of centre between consecutive legs.
    pub fn retunes(&self) -> usize {
        self.legs
            .windows(2)
            .filter(|w| w[0].center_hz != w[1].center_hz)
            .count()
    }

    /// The drawn vertices, in time order: each leg's start then its end (a dwell's vertical run),
    /// consecutive legs joined by the retune's jump.
    pub fn vertices(&self) -> Vec<TuneVertex> {
        let mut out = Vec::with_capacity(self.legs.len() * 2);
        for l in &self.legs {
            out.push(TuneVertex {
                t: l.time.start,
                f_hz: l.center_hz,
                sample_rate_hz: l.sample_rate_hz,
                at: TuneVertexAt::Start,
            });
            out.push(TuneVertex {
                t: l.time.end,
                f_hz: l.center_hz,
                sample_rate_hz: l.sample_rate_hz,
                at: TuneVertexAt::End,
            });
        }
        out
    }

    /// A stable id: the device and the run's first instant, so the same run keeps its id as a pane
    /// is panned across it (the rule `/api/paths` keeps for a traced emission).
    pub fn id(&self) -> String {
        format!(
            "tune:{}:{}",
            self.device.as_str(),
            self.legs
                .first()
                .map(|l| l.time.start.as_unix_nanos())
                .unwrap_or_default()
        )
    }
}

/// Every front end's route over `spans`, one [`TunePath`] per continuous run, ordered by device id
/// and then by time.
///
/// `spans` may come from any mix of sources and in any order; duplicates of one tune (the module
/// header's three descriptions of a dwell, and a notched window's three spans) collapse into one
/// leg.
pub fn tune_paths(spans: &[CoverageSpan], cfg: &TunePathConfig) -> Vec<TunePath> {
    let mut by_device: BTreeMap<String, (Device, Vec<TuneLeg>)> = BTreeMap::new();
    for s in spans {
        if s.time.duration_ns() <= 0 || !s.center_hz.is_finite() {
            continue;
        }
        by_device
            .entry(s.device.as_str().to_owned())
            .or_insert_with(|| (s.device.clone(), Vec::new()))
            .1
            .push(TuneLeg {
                time: s.time,
                center_hz: s.center_hz,
                sample_rate_hz: s.sample_rate_hz,
            });
    }
    let mut out = Vec::new();
    for (_, (device, legs)) in by_device {
        for run in runs(legs, cfg) {
            out.push(TunePath {
                device: device.clone(),
                legs: run.0,
                truncated: run.1,
            });
        }
    }
    out
}

/// Merges one device's raw legs into a timeline, then cuts it into continuous runs.
fn runs(mut legs: Vec<TuneLeg>, cfg: &TunePathConfig) -> Vec<(Vec<TuneLeg>, bool)> {
    legs.sort_by(|a, b| {
        a.time
            .start
            .as_unix_nanos()
            .cmp(&b.time.start.as_unix_nanos())
            .then(a.center_hz.total_cmp(&b.center_hz))
    });
    // One timeline: the same tune described by several sources is one leg, and a later leg at a
    // different centre never overlaps an earlier one (a front end has one tune at a time, so an
    // overlap is record jitter at a seal — the later leg starts where the earlier ended).
    let mut merged: Vec<TuneLeg> = Vec::with_capacity(legs.len());
    for leg in legs {
        match merged.last_mut() {
            Some(prev)
                if prev.center_hz == leg.center_hz
                    && leg.time.start.as_unix_nanos() - prev.time.end.as_unix_nanos()
                        <= cfg.join_gap_ns =>
            {
                if leg.time.end > prev.time.end {
                    prev.time = TimeRange::new(prev.time.start, leg.time.end);
                    prev.sample_rate_hz = leg.sample_rate_hz;
                }
            }
            Some(prev) if leg.time.start < prev.time.end => {
                // A different tune overlapping the previous one: clip it to start at the seal.
                let start = prev.time.end;
                if leg.time.end > start {
                    merged.push(TuneLeg {
                        time: TimeRange::new(start, leg.time.end),
                        ..leg
                    });
                }
            }
            _ => merged.push(leg),
        }
    }
    // Cut at every gap the records do not cover.
    let mut out: Vec<(Vec<TuneLeg>, bool)> = Vec::new();
    let mut run: Vec<TuneLeg> = Vec::new();
    for leg in merged {
        let gapped = run.last().is_some_and(|p: &TuneLeg| {
            leg.time.start.as_unix_nanos() - p.time.end.as_unix_nanos() > cfg.join_gap_ns
        });
        if gapped && !run.is_empty() {
            out.push(bound(std::mem::take(&mut run), cfg));
        }
        run.push(leg);
    }
    if !run.is_empty() {
        out.push(bound(run, cfg));
    }
    out
}

/// Keeps at most `max_vertices / 2` legs — the newest, the end a live view is looking at.
fn bound(mut legs: Vec<TuneLeg>, cfg: &TunePathConfig) -> (Vec<TuneLeg>, bool) {
    let max_legs = (cfg.max_vertices / 2).max(1);
    let truncated = legs.len() > max_legs;
    if truncated {
        legs.drain(..legs.len() - max_legs);
    }
    (legs, truncated)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::coverage::Analysis;
    use hk_model::FreqRange;

    fn ts(s: f64) -> Timestamp {
        Timestamp::from_unix_nanos((s * 1e9) as i64)
    }

    fn span(device: &str, t0: f64, t1: f64, center_hz: f64) -> CoverageSpan {
        CoverageSpan {
            device: if device == "unknown" {
                Device::Unknown
            } else {
                Device::Id(device.to_owned())
            },
            time: TimeRange::new(ts(t0), ts(t1)),
            freq: FreqRange::new(center_hz - 1e6, center_hz + 1e6),
            analysis: Analysis::Analysed,
            center_hz,
            sample_rate_hz: 2e6,
        }
    }

    #[test]
    fn a_dwell_is_two_vertices_and_a_retune_is_where_the_centre_changes() {
        let spans = [
            span("a", 0.0, 10.0, 100e6),
            span("a", 10.0, 20.0, 200e6),
            span("a", 20.0, 25.0, 300e6),
        ];
        let paths = tune_paths(&spans, &TunePathConfig::default());
        assert_eq!(paths.len(), 1, "one continuous run: {paths:?}");
        let p = &paths[0];
        assert_eq!(p.device, Device::Id("a".into()));
        assert_eq!(p.retunes(), 2);
        assert_eq!(p.freq_extent(), Some((100e6, 300e6)));
        let v = p.vertices();
        assert_eq!(v.len(), 6);
        // Each dwell is a vertical run: same frequency, two instants.
        assert_eq!((v[0].f_hz, v[1].f_hz), (100e6, 100e6));
        assert_eq!(v[0].at, TuneVertexAt::Start);
        assert_eq!(v[1].at, TuneVertexAt::End);
        // The retune's vertex is at the recorded instant, at the new centre.
        assert_eq!((v[2].t, v[2].f_hz), (ts(10.0), 200e6));
        assert_eq!(v[5].t, ts(25.0));
        assert!(!p.truncated);
    }

    #[test]
    fn one_tune_described_by_several_sources_is_one_leg() {
        // The ring journal, the sealed record, the open dwell and the DC notch all describe the
        // same dwell; overlapping and duplicate spans must not stack into coincident lines.
        let spans = [
            span("a", 0.0, 10.0, 100e6),
            span("a", 0.0, 10.0, 100e6),
            span("a", 4.0, 12.0, 100e6),
            span("a", 11.9, 14.0, 100e6),
        ];
        let paths = tune_paths(&spans, &TunePathConfig::default());
        assert_eq!(paths.len(), 1);
        assert_eq!(paths[0].legs.len(), 1, "{:?}", paths[0].legs);
        assert_eq!(paths[0].legs[0].time, TimeRange::new(ts(0.0), ts(14.0)));
        assert_eq!(paths[0].retunes(), 0);
    }

    #[test]
    fn a_gap_no_record_covers_breaks_the_route_rather_than_drawing_across_it() {
        let spans = [
            span("a", 0.0, 10.0, 100e6),
            span("a", 60.0, 70.0, 100e6),
            span("a", 70.0, 80.0, 200e6),
        ];
        let paths = tune_paths(&spans, &TunePathConfig::default());
        assert_eq!(paths.len(), 2, "the 50 s gap is not a leg: {paths:?}");
        assert_eq!(paths[0].legs.len(), 1);
        assert_eq!(paths[1].legs.len(), 2);
        assert_eq!(paths[0].time(), TimeRange::new(ts(0.0), ts(10.0)));
        assert_eq!(paths[1].time(), TimeRange::new(ts(60.0), ts(80.0)));
        assert_ne!(paths[0].id(), paths[1].id());
    }

    #[test]
    fn each_front_end_gets_its_own_route_and_unknown_is_its_own() {
        let spans = [
            span("a", 0.0, 10.0, 100e6),
            span("b", 0.0, 10.0, 900e6),
            span("unknown", 0.0, 10.0, 50e6),
        ];
        let paths = tune_paths(&spans, &TunePathConfig::default());
        assert_eq!(paths.len(), 3);
        let devices: Vec<&str> = paths.iter().map(|p| p.device.as_str()).collect();
        assert_eq!(devices, ["a", "b", "unknown"]);
        assert!(paths.iter().all(|p| p.legs.len() == 1));
    }

    #[test]
    fn an_overlapping_later_tune_starts_where_the_earlier_one_ended() {
        let spans = [span("a", 0.0, 10.0, 100e6), span("a", 9.0, 20.0, 200e6)];
        let paths = tune_paths(&spans, &TunePathConfig::default());
        assert_eq!(paths.len(), 1);
        assert_eq!(
            paths[0].legs.iter().map(|l| l.time).collect::<Vec<_>>(),
            [
                TimeRange::new(ts(0.0), ts(10.0)),
                TimeRange::new(ts(10.0), ts(20.0))
            ],
            "a front end has one tune at a time"
        );
    }

    #[test]
    fn a_long_route_keeps_its_newest_vertices_and_says_it_was_cut() {
        let cfg = TunePathConfig {
            max_vertices: 6,
            ..Default::default()
        };
        let spans: Vec<CoverageSpan> = (0..10)
            .map(|k| {
                let t = f64::from(k);
                span("a", t, t + 1.0, 100e6 + f64::from(k) * 1e6)
            })
            .collect();
        let paths = tune_paths(&spans, &cfg);
        assert_eq!(paths.len(), 1);
        assert!(paths[0].truncated);
        assert_eq!(paths[0].vertices().len(), 6);
        assert_eq!(paths[0].time().end, ts(10.0), "the newest end is kept");
    }
}
